segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/delegationtoken/TokenVerifierImpl.java

```diff
@@ -9,55 +9,95 @@
+            this.tokenSigningKey = tokenSigningKeyBasis.getBytes();
+        }
     }
 
     @Override
-    public boolean verifyToken(String resource, String token, AuthHandler.Permissions expectedLevel) {
-        if (config.isAuthEnabled()) {
-            try {
-                Jws<Claims> claims = Jwts.parser()
-                                         .setSigningKey(config.getTokenSigningKey().getBytes())
-                                         .parseClaimsJws(token);
-                Optional<Map.Entry<String, Object>> matchingClaim = claims.getBody().entrySet().stream().filter(entry ->
-                        validateEntry(entry, resource)
-                        && expectedLevel.compareTo(AuthHandler.Permissions.valueOf(entry.getValue().toString()))
-                        <= 0).findFirst();
-                if (matchingClaim.isPresent()) {
-                    log.debug("Found a matching claim {} for resource {}", matchingClaim, resource);
-                    return true;
-                } else {
-                    log.debug("Could not find a matching claim {} for resource {} in claims {}",
-                            expectedLevel, resource, claims);
-                    return false;
-                }
-            } catch (JwtException e) {
-                log.warn("Claim verification failed for resource {} because {}", resource, e);
-                return false;
-            }
-        } else {
+    public boolean isTokenValid(String resource, String token, AuthHandler.Permissions expectedLevel) {
+        try {
+            verifyToken(resource, token, expectedLevel);
             return true;
+        } catch (TokenException e) {
+            log.warn(e.getMessage(), e);
+            return false;
         }
     }
 
-    private boolean validateEntry(Map.Entry<String, Object> entry, String resource) {
-        return (entry.getKey().endsWith("/") && resource.startsWith(entry.getKey()))
-                    ||  resource.startsWith(entry.getKey() + "/")
-                || entry.getKey().equals("*");
+    @Override
+    public void verifyToken(String resource, String token, AuthHandler.Permissions expectedLevel)
+            throws TokenExpiredException, InvalidTokenException, InvalidClaimException, TokenException {
+        if (isAuthEnabled) {
+
+            // All key value pairs inside the payload are returned, including standard fields such as sub (for subject),
+            // aud (for audience), iat, exp, as well as custom fields of the form "<resource> -> <permission>" set by
+            // Pravega.
+            Set<Map.Entry<String, Object>> claims =
+                    JsonWebToken.fetchClaims(token, tokenSigningKey);
+
+            Optional<Map.Entry<String, Object>> matchingClaim = claims.stream()
+                    .filter(entry -> entryMatchesResource(entry, resource) &&
+                                     expectedLevel.compareTo(AuthHandler.Permissions.valueOf(entry.getValue().toString())) <= 0)
```

Comment:
Are you doing an ENUM value comparison? This enum does not have statically assigned values. If someone decides to reorder them in the code, your comparison will break. I suggest using a stronger check that would handle such a situation.

Reply:
The enum, as well as the comparison made here, is pre-existing code. I am thinking it is prudent to do this change separately in a different PR for the following reasons:

1. There might be unexpected repercussions. We also support AuthHandler plugins that use the Enum.
2. Plus there are other pending changes that were discussed in other issues/PRs that require changing the Permissions enum. 

What do you say?